main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala

```diff
@@ -463,6 +474,24 @@ object Unification {
       Ok((s, econstrs, renv, lenv, purifiedEff))
     }
 
+  // MATT docs / hack
+  def purifyEffAndRefresh(tvar: Type.Var, eff: Type)(implicit level: Level, flix: Flix): InferMonad[Type] = {
+    InferMonad { case (s, econstrs, renv, lenv) =>
+
+      val cache = mutable.HashMap.empty[Symbol.KindedTypeVarSym, Type.Var]
+
+      val res = s(eff).map {
+        // Case 1: The region. Purify it.
+        case tv if tv.sym == tvar.sym => Type.Pure
+        // Case 2: A variable from inside the region. Generalize it.
+        case tv if tv.sym.level > level => cache.getOrElseUpdate(tv.sym, Type.freshVar(tv.kind, SourceLocation.Unknown))
```

Comment:
This introduces letgen? or?

Reply:
Or not really, only kind of.

Reply:
It's effectively letgen and then immediate instantiation to free variables.

Reply:
> It's effectively letgen and then immediate instantiation to free variables.

Right. Which changes the semantics slightly, right?

I am surprised monomorphization does not fail, because it seems that types would not be propagated into a let bound lambda?

Reply:
The let-gen only happens with variables in the effect, not the type. So for types I don't think there is a monomorphization problem. But there might be an issue with purity reflection.